niceman/distributions/docker.py

```diff
@@ -0,0 +1,169 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the niceman package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Support for Docker distribution(s)."""
+
+import attr
+import json
+import logging
+
+lgr = logging.getLogger('niceman.distributions.docker')
+
+from .base import Package
+from .base import Distribution
+from .base import DistributionTracer
+from .base import TypedList
+from .base import _register_with_representer
+from ..dochelpers import borrowdoc
+from ..support.exceptions import CommandError
+
+
+@attr.s(slots=True, frozen=True, cmp=False, hash=True)
+class DockerImage(Package):
+    """Docker image information"""
+    id = attr.ib()
+    # Optional
+    architecture = attr.ib(default=None)
+    operating_system = attr.ib(default=None)
+    docker_version = attr.ib(default=None)
+    repo_digests = attr.ib(default=None)
+    repo_tags = attr.ib(default=None)
+    created = attr.ib(default=None)
+
+_register_with_representer(DockerImage)
+
+@attr.s
+class DockerDistribution(Distribution):
+    """
+    Class to provide commands to Docker engine.
+    """
+
+    images = TypedList(DockerImage)
+
+    def initiate(self, session):
+        """
+        Perform any initialization commands needed in the environment
+        environment.
+
+        Parameters
+        ----------
+        session : object
+            The Session to work in
+        """
+
+        # Raise niceman.support.exceptions.CommandError exception if Docker 
+        # engine is not to be found.
+        session.execute_command(['docker', 'info'])
+
+    def install_packages(self, session=None):
+        """
+        Install the Docker images associated to this distribution by the
+        provenance into the environment.
+
+        Parameters
+        ----------
+        session : object
+            Session to work in
+        """
+
+        for image in self.images:
+
+            # First, look in local Docker engine.
+            try:
+                session.execute_command(['docker', 'image', 'inspect',
```

Comment:
`session` is an optional argument.  If nothing is passed, wouldn't this result in an AttributeError?